db/migrate/20210212223928_handle_nil_variants.rb

```diff
@@ -0,0 +1,56 @@
+class HandleNilVariants < ActiveRecord::Migration
+   def up
+     Spree::Variant.where(product_id: nil).destroy_all
+     Spree::Variant.where(unit_value: [nil, Float::NAN]).find_each do |variant|
+       if variant.price.nil? && variant.product.master == variant
+         variant.update_column(:unit_value, 1)
+       elsif !variant.valid?
+         variant.update_column(:unit_value, 1)
+       else
+         variant.unit_value = 1
+         begin
+           variant.save
+         rescue
+           variant.update_column(:unit_value, 0)
+         end
+       end
+     end
+     Spree::Variant.where(weight: [nil, Float::NAN]).find_each do |variant|
+       if variant.price.nil? && variant.product.master == variant
+         variant.update_column(:weight, 0)
+       elsif !variant.valid?
+         variant.update_column(:weight, 0)
+       else
+         variant.weight = 0
+         begin
+           variant.save
+         rescue
+           variant.update_column(:weight, 0)
+         end
+       end
+     end
+     change_column_null :spree_variants, :unit_value, false, 1
+     change_column_null :spree_variants, :weight, false, 0.0
+     change_column_default :spree_variants, :unit_value, 1
+     change_column_default :spree_variants, :weight, 0.0
+     begin
+       execute "ALTER TABLE spree_variants ADD CONSTRAINT check_unit_value_for_nan CHECK (unit_value <> 'NaN')"
+       execute "ALTER TABLE spree_variants ADD CONSTRAINT check_weight_for_nan CHECK (weight <> 'NaN')"
+     rescue ActiveRecord::StatementInvalid
```

Comment:
are we really sure this won't catch things we didn't anticipate and which could go unnoticed? I'd also match against the exception message to be extra sure, if there's no more specific PostgreSQL exception class we can rescue.

Reply:
It turns out this won't actually work as I intended it; I wrote a helper method to check if the constraint exists instead: 05d49f76d51fd69a9ae7337bae240ca829a6736c